lib/render/vp2RenderDelegate/proxyRenderDelegate.cpp

```diff
@@ -0,0 +1,563 @@
+bool ProxyRenderDelegate::getInstancedSelectionPath(
+    const MHWRender::MRenderItem& renderItem,
+    const MHWRender::MIntersection& intersection,
+    MDagPath& dagPath) const
+{
+#if defined(WANT_UFE_BUILD)
+    // When point snapping, only the point position matters, so return false
+    // to use the DAG path from the default implementation and avoid the UFE
+    // global selection list to be updated.
+    if (pointSnappingActive())
+        return false;
+
+    auto handler = Ufe::RunTimeMgr::instance().hierarchyHandler(USD_UFE_RUNTIME_ID);
+    if (handler == nullptr)
+        return false;
+
+    MayaUsdProxyShapeBase* proxyShape = getProxyShape();
+    if (proxyShape == nullptr)
+        return false;
+
+    // Extract id of the owner Rprim. A SdfPath directly created from the render
+    // item name could be ill-formed if the render item represents instancing:
+    // "/TreePatch/Tree_1.proto_leaves_id0/DrawItem_xxxxxxxx". Thus std::string
+    // is used instead to extract Rprim id.
+    const std::string renderItemName = renderItem.name().asChar();
+    const auto pos = renderItemName.find_last_of(USD_UFE_SEPARATOR);
+    SdfPath rprimId(renderItemName.substr(0, pos));
+
+    // If the selection hit comes from an instanced render item, its instance
+    // transform matrices should have been sorted according to USD instance ID,
+    // therefore drawInstID is usdInstID plus 1 considering VP2 defines the
+    // instance ID of the first instance as 1.
+    const int drawInstID = intersection.instanceID();
+    if (drawInstID > 0) {
+        const int usdInstID = drawInstID - 1;
+        rprimId = _sceneDelegate->GetPathForInstanceIndex(rprimId, usdInstID, nullptr);
+    }
+
+    const SdfPath usdPath(_sceneDelegate->ConvertIndexPathToCachePath(rprimId));
+
+    const Ufe::PathSegment pathSegment(usdPath.GetText(), USD_UFE_RUNTIME_ID, USD_UFE_SEPARATOR);
+    const Ufe::SceneItem::Ptr& si = handler->createItem(proxyShape->ufePath() + pathSegment);
+    if (!si) {
+        TF_WARN("UFE runtime is not updated for the USD stage. Please save scene and reopen.");
+        return false;
+    }
+
+    auto globalSelection = Ufe::GlobalSelection::get();
+
+    // If update for selection is enabled, we can query selection list adjustment
+    // mode once per selection update to avoid any potential performance hit due
+    // to MEL command execution.
+#if defined(MAYA_ENABLE_UPDATE_FOR_SELECTION)
+    switch (_globalListAdjustment)
+#else
+    switch (GetListAdjustment())
+#endif
+    {
+    case MGlobal::kReplaceList:
+        // The list has been cleared before viewport selection runs, so we
+        // can add the new hits directly. UFE selection list is a superset
+        // of Maya selection list, calling clear()/replaceWith() on UFE
+        // selection list would clear Maya selection list.
+        globalSelection->append(si);
+        break;
+    case MGlobal::kAddToList:
+        globalSelection->append(si);
```

Comment:
This is only selection interpretation. Selection sorting has been done by Maya before entering this function, and only the closest hits (during camera-based selection) can get chance to invoke this function.

Reply:
I've just tested this. Single click selection is currently selecting every object under the cursor, rather than the closest one to the camera. Depth info is not being handled by Maya from what I can see?

Reply:
This is not expected, so it looks like a bug. Which platform are you using? And do you see this message in output window?

> #  VP2_RENDER_DELEGATE_PROXY is overridden to 'true'.  Default is 'false'.  #

Reply:
This is a bug in Maya 2019. To workaround it, please turn on camera-based selection. Currently we don't know whether there will be another 2019 update; if so, we will fix the issue.